### PR TITLE
Introduce UserDefinedType

### PR DESCRIPTION
**Pull request by rongrong**
Depended by https://github.com/facebookexternal/presto-facebook/pull/1372

Introduce the concept of UserDefinedType as spedified in spec16. User defined
types are managed by FunctionNamespaceManager.

The spec has more metadata specified for user defined types but we currently
only add
```
<user-defined type definition> ::= CREATE TYPE <user-defined type body>
<user-defined type body> ::= <schema-resolved user-defined type name>
    [ AS <representation> ]
<representation> ::=
    <predefined type>
  | <collection type>
  | <member list>
```
For <representation>, we use TypeSignature to represent all options above.

Enum type signature format is changed from `typename(enum:type{enum_map})` to
`enumtype(typename{enum_map})` so enum types can behave the same way as other
builtin parametric types that's fully managed by
BuiltInTypeAndFunctionNamespaceManager.

Test plan - Travis

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by rongrong** [comment]
There seems to be a slight performance regression for type signature parsing that I can't get to the bottom of:

Before the type signature format change:
```
Benchmark                                                       Mode  Cnt    Score   Error  Units
BenchmarkTypeSignatureParsing.parseBigLongEnumTypeSignature     avgt   20   60.107 ± 6.650  ms/op
BenchmarkTypeSignatureParsing.parseBigVarcharEnumTypeSignature  avgt   20  140.270 ± 4.333  ms/op
BenchmarkTypeSignatureParsing.parseRowTypeSignature             avgt   20    2.101 ± 0.196  ms/op
BenchmarkTypeSignatureParsing.parseRowTypeSignatureWithEnums    avgt   20    2.584 ± 0.473  ms/op
```
After:
```
Benchmark                                                       Mode  Cnt    Score   Error  Units
BenchmarkTypeSignatureParsing.parseBigLongEnumTypeSignature     avgt   20   63.391 ± 4.711  ms/op
BenchmarkTypeSignatureParsing.parseBigVarcharEnumTypeSignature  avgt   20  146.647 ± 7.203  ms/op
BenchmarkTypeSignatureParsing.parseRowTypeSignature             avgt   20    3.038 ± 0.331  ms/op
BenchmarkTypeSignatureParsing.parseRowTypeSignatureWithEnums    avgt   20    3.193 ± 0.356  ms/op
```

File: presto-common/src/main/java/com/facebook/presto/common/type/SemanticType.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.common.type;
+
+import com.facebook.presto.common.QualifiedObjectName;
+
+import static java.util.Objects.requireNonNull;
+
+public class SemanticType
```
Comment (daniel-ohayon) [inline presto-common/src/main/java/com/facebook/presto/common/type/SemanticType.java:20]: since it's really just a type + a name, should we call this a `NamedType` instead?

(semantic implies that it's a type that contains info about the meaning of your data, but you could have a  semantic type that is anonymous)

Comment (daniel-ohayon) [reply presto-common/src/main/java/com/facebook/presto/common/type/SemanticType.java:20]: another thought: this looks a lot like an early version of what the SQL2016 spec calls a "distinct type"
"The definition of a distinct type specifies its name and the name of its source type." (section 4.7.2)

Comment (daniel-ohayon) [reply presto-common/src/main/java/com/facebook/presto/common/type/SemanticType.java:20]: if we ever wanted to support "opaque alias types" like
`Kilogram as bigint` and `gram as bigint`
such that
```
CAST(5 as kilogram) = cast(3 as gram)
```
is a type-checking error, would this design allow for that?

Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/SemanticType.java:20]: Ya I thought about calling it NamedType instead. It kinda depends on whether we want to make this specific to external types, or change this to be something more fundamentally used in parser (`BIGINT` the physical type can also have a corresponding semantic type `BIGINT`).

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java [hunk 2]
```diff
@@ -252,7 +260,7 @@ VarcharEnumMap getVarcharEnumMap()
         Set<Integer> indices = new HashSet<>();
         Matcher enumMatcher = ENUM_PREFIX.matcher(signature);
         while (enumMatcher.find()) {
-            indices.add(enumMatcher.start());
+            indices.add(enumMatcher.end());
```
Comment (caithagoras) [inline presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java:263]: Maybe I don't have the full picture. Why is this changed from start to end?

Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java:263]: The enum signature format is changed from `typename(enum:type{enum_map})` to
`enumtype(typename{enum_map})`

**Review by caithagoras — APPROVED** [review]

File: presto-common/src/main/java/com/facebook/presto/common/type/BigintEnumType.java [hunk 3]
```diff
@@ -157,7 +162,7 @@ public boolean equals(Object o)
             return false;
         }
 
-        LongEnumType other = (LongEnumType) o;
+        BigintEnumType other = (BigintEnumType) o;
 
         return Objects.equals(getTypeSignature().getBase(), other.getTypeSignature().getBase())
                 && Objects.equals(getEnumMap(), other.getEnumMap());
```
Comment (kaikalur) [inline presto-common/src/main/java/com/facebook/presto/common/type/BigintEnumType.java:168]: Is case sensitivity taken care of here?

File: presto-common/src/main/java/com/facebook/presto/common/type/StandardTypes.java [hunk 4]
```diff
@@ -45,6 +45,8 @@
     public static final String IPPREFIX = "ipprefix";
     public static final String GEOMETRY = "Geometry";
     public static final String BING_TILE = "BingTile";
+    public static final String BIGINT_ENUM = "BigintEnum";
+    public static final String VARCHAR_ENUM = "VarcharEnum";
```
Comment (kaikalur) [inline presto-common/src/main/java/com/facebook/presto/common/type/StandardTypes.java:49]: Is this user visible? The names are a bit odd. Not sure if  enum<varchar> will work better

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java [hunk 5]
```diff
@@ -219,31 +222,36 @@ else if (c == ',') {
...
-            checkArgument(!isLongEnum, "Invalid enum map format");
+            checkArgument(!isBigintEnum, "Invalid enum map format");
             // Varchar enum values are base32-encoded so that they are case-insensitive, which is expected of TypeSigntures
             Base32 base32 = new Base32();
-            return new VarcharEnumMap(map.entrySet().stream()
-                    .collect(Collectors.toMap(Map.Entry::getKey, e -> new String(base32.decode(e.getValue().toUpperCase(ENGLISH))))));
+            return new VarcharEnumMap(
+                    typeName,
+                    map.entrySet().stream()
+                            .collect(Collectors.toMap(Map.Entry::getKey, e -> new String(base32.decode(e.getValue().toUpperCase(ENGLISH))))));
```
Comment (kaikalur) [inline presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java:254]: Why ENGLISH?

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java [hunk 6]
```diff
@@ -273,13 +281,14 @@ VarcharEnumMap getVarcharEnumMap()
     private static EnumMapParsingData parseEnumMap(String signature, int startIndex)
     {
         EnumMapParsingState state = EnumMapParsingState.EXPECT_KEY;
-        boolean isLongEnum = signature.substring(startIndex).trim().toLowerCase(ENGLISH).startsWith(LONG_ENUM_PREFIX);
-        int openBracketIndex = startIndex + (isLongEnum ? LONG_ENUM_PREFIX.length() : VARCHAR_ENUM_PREFIX.length()) + 1;
+        boolean isBigintEnum = signature.startsWith(BIGINT_ENUM_PREFIX, startIndex - BIGINT_ENUM_PREFIX.length() - 1);
```
Comment (kaikalur) [inline presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java:284]: Is the signature already case normalized?

File: presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedType.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ */
+package com.facebook.presto.common.type;
+
+import com.facebook.presto.common.QualifiedObjectName;
+
+import static java.util.Objects.requireNonNull;
+
+public class UserDefinedType
+{
+    private final QualifiedObjectName name;
```
Comment (kaikalur) [inline presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedType.java:22]: Why qualified? Can it be just a simple id?

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java [hunk 8]
```diff
@@ -199,18 +191,16 @@ public Type getType(TypeSignature signature)
...
 
         Optional<FunctionNamespaceManager<?>> functionNamespaceManager = getServingFunctionNamespaceManager(signature.getTypeSignatureBase());
         checkArgument(functionNamespaceManager.isPresent(), "Cannot find function namespace for type '%s'", signature.getBase());
-        return instantiateParametricType(signature, functionNamespaceManager.get());
+        Optional<UserDefinedType> userDefinedType = functionNamespaceManager.get().getUserDefinedType(signature.getTypeSignatureBase().getQualifiedObjectName());
+        if (!userDefinedType.isPresent()) {
+            throw new IllegalArgumentException("Unknown type " + signature);
+        }
+        checkArgument(userDefinedType.get().getPhysicalTypeSignature().getTypeSignatureBase().isStandardType(), "UserDefinedType must be based on static types.");
+        return getType(userDefinedType.get().getPhysicalTypeSignature());
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java:203]: What is a "static type"?

**Review by kaikalur — COMMENTED** [review]
Can we split it into two PRs (not just commits) - one for UserDefined type that's of more general interest to the community and enum separately.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java (hunk 8 above)
Comment (rongrong) [reply presto-main/src/main/java/com/facebook/presto/metadata/FunctionAndTypeManager.java:203]: Built in types / types that's defined in code thus cannot be changed. Do you have a better name for them?

File: presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedType.java (hunk 7 above)
Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/UserDefinedType.java:22]: No, all external types are managed by `FunctionNamespaceManager`, which manages catalogs. This is also aligned with the spec as `<schema-resolved user-defined type name>`.

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java (hunk 6 above)
Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java:284]: Yes, the caller normalized the signature before calling. There are tests cases to test this as well.

File: presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java (hunk 5 above)
Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/TypeSignature.java:254]: I don't know... That's what's used in code in similar situations. It's no longer recommended to not providing a locale.

File: presto-common/src/main/java/com/facebook/presto/common/type/StandardTypes.java (hunk 4 above)
Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/StandardTypes.java:49]: There are 2 types of enums and they cannot be both of type "enum". This would be visible in error messages and plans. If you have better suggestions I'm all ears.

File: presto-common/src/main/java/com/facebook/presto/common/type/BigintEnumType.java (hunk 3 above)
Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/BigintEnumType.java:168]: Hmm, the enum mapping should be case sensitive. The type name probably should not.

**Comment by rongrong** [comment]
> Can we split it into two PRs (not just commits) - one for UserDefined type that's of more general interest to the community and enum separately.

I tried. The problem is either the commit / PR that introduces user defined type is not really used (only a simple container class, and dead code), or I need to change enum to use it, which requires the refactor of enum.

File: presto-common/src/main/java/com/facebook/presto/common/type/LongEnumType.java [hunk 9]
```diff
@@ -114,25 +123,22 @@ public boolean equals(Object o)
...
+            return Objects.equals(typeName, other.typeName) && Objects.equals(enumMap, other.enumMap);
         }
 
         @Override
         public String toString()
         {
-            return "enum:bigint{"
-                    + enumMap.entrySet()
-                    .stream()
+            return format("%s{%s}", typeName, enumMap.entrySet().stream()
```
Comment (kaikalur) [inline presto-common/src/main/java/com/facebook/presto/common/type/LongEnumType.java:132]: Is the order the same as the user given? That might be good.

Comment (rongrong) [reply presto-common/src/main/java/com/facebook/presto/common/type/LongEnumType.java:132]: Depending on how it's stored at the source it might or might not be possible to have the order as "user given". Since we are implementing it as map, that's not really possible. I don't want to make this like a map block that's too complicated. 😂

**Review by kaikalur — APPROVED** [review]
LGTM
